node/device/samples/dmpatterns_fwupdate_device.js

```
@@ -0,0 +1,158 @@
+// the multi-phase firmware update sequence.
+function onFirmwareUpdate(request, response) {
+  // Get the firmware image Uri from the body of the method request
+  var fwPackageUri = request.payload.fwPackageUri;
+  var result = url.parse(fwPackageUri);
+  
+  // Ensure that the url is to a secure url
+  if (result.protocol != 'https:') {
+    response.send(400, 'Invalid URL format.  Must use https:// protocol.', function(err) {
+      if (err) console.error('Error sending method response :\n' + err.toString());
+      else console.log('Response to method \'' + request.methodName + '\' sent successfully.');
+    });
+  } else {
+    // Respond the cloud app for the device method
+    response.send(200, 'Firmware update started.', function(err) {
+      if (err) console.error('Error sending method response :\n' + err.toString());
+      else console.log('Response to method \'' + request.methodName + '\' sent successfully.');
+    });
+
+    initiateFirmwareUpdateFlow(fwPackageUri, function(err){
+      if (!err) console.log("Completed firmwareUpdate flow");
+    });
+  }
+}
+
+// Implementation of firmwareUpdate flow
+function initiateFirmwareUpdateFlow(fwPackageUri, callback) {
+  async.waterfall([
+    function (callback) {
+      downloadImage(fwPackageUri, callback);
+    },
+    function (imageData, callback) {
+      applyImage(imageData, callback);
+    }
+  ], function(err) {
+    if (err) console.error('Error : ' + err.message);
+    callback(null);
+  });
+}
+
+// Helper function to update the twin reported properties.
+// Used by every phase of the firmware update.
+function reportFWUpdateThroughTwin(firmwareUpdateValue, callback) {
+  var patch = {
+      iothubDM : {
+        firmwareUpdate : firmwareUpdateValue
+      }
+  };
+  console.log(JSON.stringify(patch, null, 2));
+  client.getTwin(function(err, twin) {
+    if (!err) {
+      twin.properties.reported.update(patch, function(err) {
+        callback(null);
```

>callback(null); [](start = 8, length = 15)

swallows the error.  You could do callback(err) here, or you could skip the anonymous function and just do twin.properties.reported.udpate(patch, callback).